api/src/main/java/io/neba/api/resourcemodels/AnnotatedFieldMapper.java

```diff
@@ -0,0 +1,162 @@
+ *     It is insufficient to return a type compatible to the {@link #map(AnnotatedFieldMapper.OngoingMapping) mapping methods}
+ *     return type declaration. This return type only represents the type any returned value must be compatible to.<br />
+ *     For instance, if a mapper is responsible for {@link java.util.Collection}, it must take care to return the field's actual
+ *     collection type, e.g. {@link java.util.List} or {@link java.util.Set}. Otherwise, an exception will arise.
+ * </p>
+ *
+ * <p>
+ *     Implementations <strong>must never</strong> store any contextual data provided by the
+ *     {@link io.neba.api.resourcemodels.AnnotatedFieldMapper.OngoingMapping}
+ *     as this data stems from arbitrary OSGi bundles with independent life cycles.
+ *     Storing any data would result in a class loader / memory leak when these bundles change.
+ * </p>
+ *
+ * @param <FieldType> the super type of the {@link java.lang.reflect.Field#getType() field type}
+ *        of the mapped type.
+ * @param <AnnotationType> the exact type of the annotation this mapper is responsible for.
+
+ * @author Olaf Otto
+ */
+public interface AnnotatedFieldMapper<FieldType, AnnotationType extends Annotation> {
+    /**
+     * Represents the contextual data of a field mapping during a resource to model mapping.
+     *
+     * @author Olaf Otto
+     */
+    public interface OngoingMapping<FieldType, AnnotationType> {
+        /**
+         * @return The currently resolved value of the field,
+         * or <code>null</code> if no value could be resolved for the field.
+         */
+        FieldType getResolvedValue();
+
+        /**
+         * @return The instance of {@link #getAnnotationType() the annotation this mapper is registered for}.
+         *         Never <code>null</code>.
+         */
+        AnnotationType getAnnotation();
+
+        /**
+         * @return The mapped model. At this point, the mapping is still incomplete and
+         *         no post-processors have been invoked on the model. Never <code>null</code>.
+         */
+        Object getModel();
+
+        /**
+         * @return the mapped field. Never <code>null</code>.
+         */
+        Field getField();
+
+        /**
+         * @return All annotations (including meta-annotations, i.e. annotations of annotations) present on the field.
+         *         Never <code>null</code>.
+         */
+        Map<Class<? extends Annotation>, Annotation> getAnnotationsOfField();
+
+        /**
+         * @return The mapped type of the field. <em>Note: In case the field is {@link io.neba.api.resourcemodels.Optional}, this
+         *         type is the component type, i.e. the type targeted by the optional field</em>. However, field mappers
+         *         are not applied optional fields but to the subsequent mapping, when the {@link Optional#get() optional value is actually mapped.
```

Comment:
So you we even have to mention this then? Actually a mapper never has to deal with Optional, am I right? If so, this comment may rather be confusing.